digital-twin/device/src/main/java/com/microsoft/azure/sdk/iot/digitaltwin/device/DeviceClientManager.java

```diff
@@ -0,0 +1,164 @@
+import static com.microsoft.azure.sdk.iot.device.IotHubConnectionStatusChangeReason.RETRY_EXPIRED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED;
+import static com.microsoft.azure.sdk.iot.device.transport.IotHubConnectionStatus.DISCONNECTED_RETRYING;
+
+@Slf4j
+public class DeviceClientManager implements IotHubConnectionStatusChangeCallback {
+    private enum ConnectionStatus {
+        DISCONNECTED, CONNECTING, CONNECTED
+    }
+
+    private static final Object lock = new Object();
+    private static final int SLEEP_TIME_BEFORE_RECONNECTING_IN_SECONDS = 10;
+    private ConnectionStatus connectionStatus;
+    private Pair<IotHubConnectionStatusChangeCallback, Object> suppliedConnectionStatusChangeCallback;
+
+    private interface DeviceClientNonDelegatedFunction {
+        void open();
+        void closeNow();
+        void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext);
+    }
+
+    // The methods defined in the interface DeviceClientNonDelegatedFunction will be called on DeviceClientManager, and not on DeviceClient.
+    @Delegate(excludes = DeviceClientNonDelegatedFunction.class)
+    private final DeviceClient client;
+
+    DeviceClientManager(DeviceClient deviceClient) {
+        this.connectionStatus = ConnectionStatus.DISCONNECTED;
+        this.client = deviceClient;
+        this.client.registerConnectionStatusChangeCallback(this, this);
+    }
+
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext) {
+        this.suppliedConnectionStatusChangeCallback = new Pair<>(callback, callbackContext);
+    }
+
+    public void open() throws Exception {
+        synchronized (lock) {
+            if (connectionStatus == ConnectionStatus.DISCONNECTED) {
+                connectionStatus = ConnectionStatus.CONNECTING;
+            } else {
+                return;
+            }
+        }
+        doConnect();
+    }
+
+    private void doConnect() throws Exception {
+        // Device client does not have retry on the initial open() call. Will need to be re-opened by the calling application
+        while (connectionStatus == ConnectionStatus.CONNECTING) {
+            synchronized (lock) {
+                if (connectionStatus == ConnectionStatus.CONNECTING) {
+                    try {
+                        log.debug("[connect] - Opening the device client instance...");
+                        client.open();
+                        connectionStatus = ConnectionStatus.CONNECTED;
+                        break;
+                    }
+                    catch (Exception ex) {
+                        log.error("[connect] - Exception thrown while opening DeviceClient instance: ", ex);
+                        if (ex instanceof TransportException && ((TransportException) ex).isRetryable()) {
```

Comment:
From my testing, we throw `IOException` on network loss. The `TransportException` is embedded under `Exception.getCause()`: https://github.com/Azure/azure-iot-sdk-java/pull/681/files#diff-d05304fffaae6bfdbe785b51eca97bfcR74

Reply:
https://github.com/Azure/azure-iot-sdk-java/blob/master/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java#L162

Reply:
Too bad to hear that...